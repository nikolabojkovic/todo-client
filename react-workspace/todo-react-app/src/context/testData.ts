import { 
  ITodo,
  IPaging, 
  ISort, 
  SortDirection, 
  IFilter,
  StateFilter, 
  IGeneralSettings, 
  IPaginationSettings, 
  ISearchSettings, 
  ISettings, 
  ListContainerType, 
  PaginationType, 
  BackgroundColor,
  IThemeSettings,
  BsThemes
} from "../models";
import { DisplayMode, IState } from "./";

export const todoListTestData: ITodo[] = [
  {
    id: 1,
    title: "Task 1",
    description: "Description 1",
    completed: false,
    createdAt: new Date(2022, 1, 4)
  } as ITodo,
  {
    id: 2,
    title: "Task 2",
    description: "Description 2",
    completed: true,
    createdAt: new Date(2022, 1, 5)
  } as ITodo,
  {
    id: 3,
    title: "Task 3",
    description: "Description 3",
    completed: false,
    createdAt: new Date(2022, 1, 6)
  } as ITodo,
  {
    id: 4,
    title: "Task 4",
    description: "Description 4",
    completed: false,
    createdAt: new Date(2022, 1, 7)
  } as ITodo,
  {
    id: 6,
    title: "Task 6",
    description: "Description 6",
    completed: false,
    createdAt: new Date(2022, 2, 4)
  } as ITodo,
  {
    id: 5,
    title: "Task 5",
    description: "Description 5",
    completed: false,
    createdAt: new Date(2022, 3, 8)
  } as ITodo
] as ITodo[];

export const stateTestData: IState = 
{
  originalList: todoListTestData,
  displayList: todoListTestData,
  displayMode: DisplayMode.All,
  search: { searchTerm: '' },
  filter: { state: StateFilter.all } as IFilter,
  sort: { column: 'createdAt', direction: SortDirection.Asc} as ISort,
  paging: {
    totalCount: 6,
    activePage: 1,
    itemsPerPage: 5,
    startIndex: 0,
    endIndex: 5,
  } as IPaging,
  settings: {
    general: {
      isConfirmEnabled: true,
      isPaginationEnabled: true,
      isInfiniteScrollEnabled: false,
      listSizeType: ListContainerType.Dynamic,
      fixedListSize: 200
    } as IGeneralSettings,
    search: {
      isSearchOnKeyPressEnabled: false,
      debounceTime: 500
    } as ISearchSettings,
    pagination: {
      paginationType: PaginationType.Classic,
      maxVisiblePages: 3
    } as IPaginationSettings,
    theme: {
      backgroundColor: BackgroundColor.DarkGray,
      primaryColor: '#ff9900',
      bsTheme: BsThemes.Light,
      primaryColorTopCord: -35,
      primaryColorLefCord: 81
    } as IThemeSettings
  } as ISettings
} as IState;