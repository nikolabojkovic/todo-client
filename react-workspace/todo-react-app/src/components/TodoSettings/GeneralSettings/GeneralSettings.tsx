import { Dropdown, Form, Stack } from "react-bootstrap";

import { useTodoList, useTodoListDispatch } from "../../../context";
import { IAction, TodoActions, IGeneralSettings, ListContainerType } from "../../../models";

export type Props = {
  paginationSwitch: boolean;
  onPaginationSwitch: (enabled: boolean) => void
}

export function GeneralSettings() {
  const todoList = useTodoList();
  const dispatch = useTodoListDispatch();

  function handleSettingsUpdate(generalSettingsPayload: IGeneralSettings) {
    dispatch({
      type: TodoActions.settingsUpdated,
      payload: {
        ...todoList.settings,
        general: {...generalSettingsPayload}
      }
    } as IAction);
  }
  
  function handleListSizeSelect(otpion: ListContainerType) {
    handleSettingsUpdate({
      ...todoList.settings.general,
      listSizeType: otpion
    });
  }

  return (
    <div className='App__settings__group'>
      <label className='App__settings__group-label'>General</label>
      <div className='App__settings__group__item'>
        <Stack direction="horizontal" gap={2}>
          <label>Comfirm on change/remove</label>
          <Form.Check 
            className='ms-auto'
            type="switch"
            id="comfirm-enabled"
            data-testid="comfirm-enabled"
            checked={todoList.settings.general.isConfirmEnabled}
            onChange={(e) => {
              handleSettingsUpdate({
                ...todoList.settings.general,
                isConfirmEnabled: e.target.checked
              });
            }}
          /> 
        </Stack>
      </div>
      <div className='App__settings__group__item'>
        <Stack direction="horizontal" gap={2}>
          <label>Pagination</label>
          <Form.Check 
            className='ms-auto'
            type="switch"
            id="pagination-enabled"
            data-testid="pagination-enabled"
            checked={todoList.settings.general.isPaginationEnabled}
            onChange={(e) => {
              handleSettingsUpdate({
                ...todoList.settings.general,
                isPaginationEnabled: e.target.checked,
                isInfiniteScrollEnabled: false
              });
            }}
          /> 
        </Stack>
      </div>
      <div className='App__settings__group__item'>
        <Stack direction="horizontal" gap={2}>
          <label>Infinite Scroll</label>
          <Form.Check 
            className='ms-auto'
            type="switch"
            id="infinite-scroll-enabled"
            data-testid="infinite-scroll-enabled"
            checked={todoList.settings.general.isInfiniteScrollEnabled}
            onChange={(e) => {
              handleSettingsUpdate({
                ...todoList.settings.general,
                isPaginationEnabled: false,
                isInfiniteScrollEnabled: e.target.checked
              });
            }}
          /> 
        </Stack>
      </div>
      <div className='App__settings__group__item'>
        <Stack direction="horizontal" gap={2}>
          <label>List container size</label>
          <Dropdown className="ms-auto">
            <Dropdown.Toggle
              data-testid="list-size-dropdown"
              variant="outline-secondary"
              className="action-button ps-2 pe-2"
              id="dropdown-list-size-type"
              disabled={false}
            >
              {' '}{todoList.settings.general.listSizeType}{' '}
            </Dropdown.Toggle>
            <Dropdown.Menu>
              <Dropdown.Item 
                data-testid="list-size-option-fixed" 
                onClick={() => handleListSizeSelect(ListContainerType.Fixed)}
              >
                { ListContainerType.Fixed }
              </Dropdown.Item>
              <Dropdown.Item 
                data-testid="list-size-option-dynamic" 
                onClick={() => handleListSizeSelect(ListContainerType.Dynamic)}
                >
                  { ListContainerType.Dynamic }
                </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Stack>
      </div>
      { todoList.settings.general.listSizeType === ListContainerType.Fixed &&
        <div className='App__settings__group__item ms-5'>
          <Stack direction="horizontal" gap={2}>
            <label>Size (px)</label>
            <Form.Control
              className="ms-auto"
              data-testid="list-size-input"
              type="number" 
              placeholder={'0'} 
              id="fixed-size"
              size="sm" 
              value={todoList.settings.general.fixedListSize}
              onChange={(e) => {
                handleSettingsUpdate({
                  ...todoList.settings.general,
                  fixedListSize: (+e.target.value)
                });
              }}
            />
          </Stack>
        </div>
      }      
    </div>
  );
}