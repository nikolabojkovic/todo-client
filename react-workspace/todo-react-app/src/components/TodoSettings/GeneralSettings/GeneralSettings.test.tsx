import renderer, { act } from 'react-test-renderer';
import { fireEvent, render, waitFor, screen } from "@testing-library/react";

import { IState, TodosContext, TodosDispatchContext, stateTestData } from "../../../context";
import { GeneralSettings } from "./GeneralSettings";
import { IAction, IGeneralSettings, ISettings, ListContainerType, TodoActions } from '../../../models';
import userEvent from '@testing-library/user-event';

describe('GeneralSettings', () => {
  const globalContext = {
    state: {
      ...stateTestData,
    },
    dispatch: jest.fn()
  };
  
  it('match snapshot general settings', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false 
      }
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <GeneralSettings></GeneralSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot general settings list size type fixed', () => {
    const context = {
      ...globalContext,
      state: { 
        ...globalContext.state,
        isLoading: false,
        settings: {
          ...globalContext.state.settings,
          general: {
            ...globalContext.state.settings.general,
            listSizeType: ListContainerType.Fixed
          } as IGeneralSettings
        } as ISettings
      } as IState
    };
    const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
          <GeneralSettings></GeneralSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
    const tree = renderer.create(
      jsxElement
    ).toJSON();
    render(jsxElement);
    expect(tree).toMatchSnapshot();
  });

  describe('actions', () => {
    it('should update settings on confirm on change/delete swich toggle', async () => {
      const context = {
        state: {
          ...stateTestData,
          search: { searchTerm: '' },
          settings: { 
            ...stateTestData.settings,
            search: {
              ...stateTestData.settings.search,
              isSearchOnKeyPressEnabled: true
            }
          }
        },
        dispatch: jest.fn()
      };
      const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <GeneralSettings></GeneralSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
      await act( async () => {
        render(jsxElement);
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            isConfirmEnabled: false
          }
        }
      } as IAction;
  
      const toggle = screen.getByTestId('comfirm-enabled');
      fireEvent.click(toggle);
   
      await waitFor(() => {
        expect(context.dispatch).toBeCalledWith(action);
      }); 
    });

    it('should update settings on confirm on pagination enabled swich toggle', async () => {
      const context = {
        state: {
          ...stateTestData,
          search: { searchTerm: '' },
          settings: { 
            ...stateTestData.settings,
            search: {
              ...stateTestData.settings.search,
              isSearchOnKeyPressEnabled: true
            }
          }
        },
        dispatch: jest.fn()
      };
      const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <GeneralSettings></GeneralSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
      await act( async () => {
        render(jsxElement);
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            isPaginationEnabled: false
          }
        }
      } as IAction;
  
      const toggle = screen.getByTestId('pagination-enabled');
      fireEvent.click(toggle);
   
      await waitFor(() => {
        expect(context.dispatch).toBeCalledWith(action);
      }); 
    });

    it('should update settings on confirm on infinite scroll enabled swich toggle', async () => {
      const context = {
        state: {
          ...stateTestData,
          search: { searchTerm: '' },
          settings: { 
            ...stateTestData.settings,
            search: {
              ...stateTestData.settings.search,
              isSearchOnKeyPressEnabled: true
            }
          }
        },
        dispatch: jest.fn()
      };
      const jsxElement = 
      (<TodosContext.Provider value={context.state}>
         <TodosDispatchContext.Provider value={context.dispatch} >
            <GeneralSettings></GeneralSettings>
         </TodosDispatchContext.Provider>
       </TodosContext.Provider>);
      await act( async () => {
        render(jsxElement);
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            isInfiniteScrollEnabled: true,
            isPaginationEnabled: false
          } as IGeneralSettings
        }
      } as IAction;
  
      const toggle = screen.getByTestId('infinite-scroll-enabled');
      fireEvent.click(toggle);
   
      await waitFor(() => {
        expect(context.dispatch).toBeCalledWith(action);
      }); 
    });

    it('should update settings on list size type fixed option selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              listSizeType: ListContainerType.Dynamic
            } as IGeneralSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <GeneralSettings></GeneralSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const listSizeDropdown = screen.getByTestId('list-size-dropdown');
  
      await act(async () => {      
        return await userEvent.click(listSizeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            listSizeType: ListContainerType.Fixed
          } as IGeneralSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('list-size-option-fixed');
        
      fireEvent.click(dropdownOption);  
   
      expect(context.dispatch).toBeCalledWith(action);
    });

    it('should update settings on list size type dynamic option selected', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              listSizeType: ListContainerType.Fixed
            } as IGeneralSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <GeneralSettings></GeneralSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
      const listSizeDropdown = screen.getByTestId('list-size-dropdown');
  
      await act(async () => {      
        return await userEvent.click(listSizeDropdown);    
      });
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            listSizeType: ListContainerType.Dynamic
          } as IGeneralSettings
        }
      } as IAction;
      const dropdownOption = screen.getByTestId('list-size-option-dynamic');
        
      fireEvent.click(dropdownOption);  
     
      expect(context.dispatch).toBeCalledWith(action);
    });

    it('should update settings on list size input changed', async () => {
      const context = {
        ...globalContext,
        state: { 
          ...globalContext.state,
          isLoading: false,
          settings: {
            ...globalContext.state.settings,
            general: {
              ...globalContext.state.settings.general,
              listSizeType: ListContainerType.Fixed
            } as IGeneralSettings
          } as ISettings
        } as IState
      };
      render(
        (
          <TodosContext.Provider value={context.state}>
            <TodosDispatchContext.Provider value={context.dispatch} >
              <GeneralSettings></GeneralSettings>
            </TodosDispatchContext.Provider>
          </TodosContext.Provider>
        )
      ); 
  
      const action = {
        type: TodoActions.settingsUpdated,
        payload: {
          ...context.state.settings,
          general: {
            ...context.state.settings.general,
            fixedListSize: 600
          } as IGeneralSettings
        }
      } as IAction;
        
      const listSizeInput = screen.getByTestId('list-size-input');
      fireEvent.change(listSizeInput, {target: {value: 600}});
     
      expect(context.dispatch).toBeCalledWith(action);
    });
  });
});