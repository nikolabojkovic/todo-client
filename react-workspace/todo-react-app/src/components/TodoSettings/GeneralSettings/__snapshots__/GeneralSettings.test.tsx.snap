// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralSettings match snapshot general settings 1`] = `
<div
  className="App__settings__group"
>
  <label
    className="App__settings__group-label"
  >
    General
  </label>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Comfirm on change/remove
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={true}
          className="form-check-input"
          data-testid="comfirm-enabled"
          disabled={false}
          id="comfirm-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Pagination
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={true}
          className="form-check-input"
          data-testid="pagination-enabled"
          disabled={false}
          id="pagination-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Infinite Scroll
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={false}
          className="form-check-input"
          data-testid="infinite-scroll-enabled"
          disabled={false}
          id="infinite-scroll-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        List container size
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="list-size-dropdown"
          disabled={false}
          id="dropdown-list-size-type"
          onClick={[Function]}
          type="button"
        >
           
          Dynamic
           
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`GeneralSettings match snapshot general settings list size type fixed 1`] = `
<div
  className="App__settings__group"
>
  <label
    className="App__settings__group-label"
  >
    General
  </label>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Comfirm on change/remove
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={true}
          className="form-check-input"
          data-testid="comfirm-enabled"
          disabled={false}
          id="comfirm-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Pagination
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={true}
          className="form-check-input"
          data-testid="pagination-enabled"
          disabled={false}
          id="pagination-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Infinite Scroll
      </label>
      <div
        className="ms-auto form-switch"
      >
        <input
          checked={false}
          className="form-check-input"
          data-testid="infinite-scroll-enabled"
          disabled={false}
          id="infinite-scroll-enabled"
          onChange={[Function]}
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        List container size
      </label>
      <div
        className="ms-auto dropdown"
      >
        <button
          aria-expanded={false}
          className="action-button ps-2 pe-2 dropdown-toggle btn btn-outline-secondary"
          data-testid="list-size-dropdown"
          disabled={false}
          id="dropdown-list-size-type"
          onClick={[Function]}
          type="button"
        >
           
          Fixed
           
        </button>
      </div>
    </div>
  </div>
  <div
    className="App__settings__group__item ms-5"
  >
    <div
      className="hstack gap-2"
    >
      <label>
        Size (px)
      </label>
      <input
        className="ms-auto form-control form-control-sm"
        data-testid="list-size-input"
        id="fixed-size"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value={200}
      />
    </div>
  </div>
</div>
`;
